mindmeld/components/entity_resolver.py

```diff
@@ -42,60 +50,664 @@
+            }
+
+    new usage
+        >>> er_config = {
+                "model_type": "resolver",
+                "model_settings": {
+                    "resolver_type": "text_relevance"
+                    ...
+                }
+            }
+    """
+
+    if not er_config.get("model_settings", {}).get("resolver_type", None):
+        model_type = er_config.get("model_type")
+        if model_type == "resolver":
+            raise Exception("Could not find `resolver_type` in `model_settings` of entity resolver")
+        else:
+            logger.warning("DeprecationWarning: Use latest format of configs for entity resolver. "
+                           "See https://www.mindmeld.com/docs/userguide/entity_resolver.html "
+                           "for more details.")
+            er_config = copy.deepcopy(er_config)
+            model_settings = er_config.get("model_settings", {})
+            model_settings.update({"resolver_type": model_type})
+            er_config["model_settings"] = model_settings
+            er_config["model_type"] = "resolver"
+
+    return er_config
+
+
+def _torch(op, *args, sub="", **kwargs):
+    return _getattr(f"torch{'.' + sub if sub else ''}", op)(*args, **kwargs)
+
+
+class BertEmbedder:
+    """
+    Encoder class for bert models based on https://github.com/UKPLab/sentence-transformers
+    """
+
+    # class variable to cache bert model(s);
+    #   helps to mitigate keeping duplicate sets of large weight matrices
+    CACHE_MODELS = {}
+
+    @staticmethod
+    def _batch_to_device(batch, target_device):
+        """
+        send a pytorch batch to a device (CPU/GPU)
+        """
+        tensor = _getattr("torch", "Tensor")
+        for key in batch:
+            if isinstance(batch[key], tensor):
+                batch[key] = batch[key].to(target_device)
+        return batch
+
+    @staticmethod
+    def _num_layers(model):
+        """
+        Finds the number of layers in a given transformers model
+        """
+
+        if hasattr(model, "n_layers"):  # eg. xlm
+            num_layers = model.n_layers
+        elif hasattr(model, "layer"):  # eg. xlnet
+            num_layers = len(model.layer)
+        elif hasattr(model, "encoder"):  # eg. bert
+            num_layers = len(model.encoder.layer)
+        elif hasattr(model, "transformer"):  # eg. sentence_transformers models
+            num_layers = len(model.transformer.layer)
+        else:
+            raise ValueError(f"Not supported model {model} to obtain number of layers")
+
+        return num_layers
+
+    @staticmethod
+    def _text_length(text):
+        """
+        Help function to get the length for the input text. Text can be either
+        a list of ints (which means a single text as input), or a tuple of list of ints
+        (representing several text inputs to the model).
+
+        Union[List[int], List[List[int]]]
+        """
+        if isinstance(text, dict):
+            return len(next(iter(text.values())))
```

Comment:
Why do we pick up the length of only the first dict?

Reply:
I actually re-used this method from sentence-transformers where it was necessary to deal with different kinds of inputs. In our use case, I think the `text` argument will always be a list of strings and I can probably remove this method altogether.